tests/Mocks/StorageEngineMock.cpp

```diff
@@ -432,6 +433,456 @@ struct IndexFactoryMock : arangodb::IndexFactory {
+    for (size_t i = 0; i < _fields.size(); ++i) {
+      auto slice = doc;
+      auto isExpansion = false;
+      for (auto fieldIt = _fields[i].begin(); fieldIt != _fields[i].end(); ++fieldIt) {
+        TRI_ASSERT(slice.isObject() || slice.isArray());
+        if (slice.isObject()) {
+          slice = slice.get(fieldIt->name);
+          if ((fieldIt->shouldExpand && slice.isObject()) ||
+              (!fieldIt->shouldExpand && slice.isArray())) {
+            slice = VPackSlice::nullSlice();
+            break;
+          }
+          if (slice.isNone() || slice.isNull()) {
+            break;
+          }
+        } else { // expansion
+          isExpansion = slice.isArray();
+          TRI_ASSERT(isExpansion);
+          auto found = false;
+          for (auto sliceIt = arangodb::velocypack::ArrayIterator(slice); sliceIt != sliceIt.end(); ++sliceIt) {
+            auto subSlice = sliceIt.value();
+            if (!(subSlice.isNone() || subSlice.isNull())) {
+              for (auto fieldItForArray = fieldIt; fieldItForArray != _fields[i].end(); ++fieldItForArray) {
+                TRI_ASSERT(subSlice.isObject());
+                subSlice = subSlice.get(fieldItForArray->name);
+                if (subSlice.isNone() || subSlice.isNull()) {
+                  break;
+                }
+              }
+              if (!(subSlice.isNone() || subSlice.isNull())) {
+                insertSlice(documentId, subSlice, i);
+                builder.add(subSlice);
+                found = true;
+                break;
+              }
+            }
+          }
+          if (!found) {
+            insertSlice(documentId, VPackSlice::nullSlice(), i);
+            builder.add(VPackSlice::nullSlice());
+          }
+          break;
+        }
+      }
+      if (!isExpansion) {
+        // if the last expansion (at the end) leave the array open
+        if (slice.isArray() && i == _fields.size() - 1) {
+          auto found = false;
+          auto wasNull = false;
+          for (auto sliceIt = arangodb::velocypack::ArrayIterator(slice); sliceIt != sliceIt.end(); ++sliceIt) {
+            auto subSlice = sliceIt.value();
+            if (!(subSlice.isNone() || subSlice.isNull())) {
+              insertSlice(documentId, subSlice, i);
+              found = true;
+            } else {
+              wasNull = true;
+            }
+          }
+          if (!found || wasNull) {
+            insertSlice(documentId, VPackSlice::nullSlice(), i);
+          }
+          toClose = false;
+        } else { // object
+          insertSlice(documentId, slice, i);
+          builder.add(slice);
+        }
+      }
+    }
+    if (toClose) {
+      builder.close();
+    }
+    _docIndexMap.try_emplace(documentId, std::move(builder));
+  }
+
+  bool remove(arangodb::LocalDocumentId const& documentId, arangodb::velocypack::Slice const& doc) {
+    size_t i = 0;
+    auto documentRemoved = false;
+    for (auto map : _valueMaps) {
```

Comment:
This will create a temporary copy of each vector entry, so it does not work on the original value
```suggestion
    for (auto& map : _valueMaps) {
```